staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset_test.go

```diff
@@ -949,6 +950,113 @@ func TestSelectQueueLocked(t *testing.T) {
 	}
 }
 
+func TestFinishRequestLocked(t *testing.T) {
+	tests := []struct {
+		name         string
+		workEstimate fcrequest.WorkEstimate
+	}{
+		{
+			name: "request has additional latency",
+			workEstimate: fcrequest.WorkEstimate{
+				Seats:             10,
+				AdditionalLatency: time.Minute,
+			},
+		},
+		{
+			name: "request has no additional latency",
+			workEstimate: fcrequest.WorkEstimate{
+				Seats: 10,
+			},
+		},
+	}
+
+	metrics.Register()
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			metrics.Reset()
+
+			now := time.Now()
+			clk, _ := fqclocktest.NewFakeEventClock(now, 0, nil)
+			wrappedClock := &wrappedClock{
+				EventClock: clk,
+				// the channel is not buffered intentionally, the test will time out
+				// and hence fail if finishRequestLocked does not spin up a new goroutine
+				// where the seats are held for the additional latency.
+				// this ensures that additional latency has no impact on user experience.
+				sleepDurationCh: make(chan time.Duration),
+			}
+
+			qs := &queueSet{
+				clock:   wrappedClock,
+				obsPair: newObserverPair(clk),
+			}
+			queue := &queue{}
+			r := &request{
+				qs:           qs,
+				queue:        queue,
+				workEstimate: test.workEstimate,
+			}
+
+			qs.totRequestsExecuting = 111
+			qs.totSeatsInUse = 222
+			queue.requestsExecuting = 11
+			queue.seatsInUse = 22
+
+			var (
+				queuesetTotalRequestsExecutingExpected = qs.totRequestsExecuting - 1
+				queuesetTotalSeatsInUseExpected        = qs.totSeatsInUse - int(test.workEstimate.Seats)
+				queueRequestsExecutingExpected         = queue.requestsExecuting - 1
+				queueSeatsInUseExpected                = queue.seatsInUse - int(test.workEstimate.Seats)
+			)
+
+			qs.finishRequestLocked(r)
+
+			// if the request has additional latency then we must ensure that
+			// Sleep has been invoked appropriately.
+			if additionalLatencyExpected := test.workEstimate.AdditionalLatency; additionalLatencyExpected > 0 {
+				select {
+				case additionalLatencyGot := <-wrappedClock.sleepDurationCh:
+					if test.workEstimate.AdditionalLatency != additionalLatencyGot {
+						t.Errorf("Expected addition latency of: %s, but got: %s", test.workEstimate.AdditionalLatency, additionalLatencyGot)
+					}
+				case <-time.After(100 * time.Millisecond):
```

Comment:
100ms is definitely too short in starved prow environment. Please use waitTimeForever (or sth like that we have).